docs/architecture/adr-0012-spark-execution-environment.md

```diff
@@ -0,0 +1,42 @@
+# 0012 - Spark Execution Environment
+
+## Context
+
+Part of the Raster Foundry processing workflow requires a source imagery ingest into a GeoTrellis Catalog. GeoTrellis depends on Apache Spark to parallelize work during an ingest, and that work typically occurs on an Apache Spark enabled compute cluster.
+
+Prior iterations of Raster Foundry attempted ingests with an Apache YARN managed compute cluster via Amazon Elastic MapReduce (EMR) on a per-request basis. Unfortunately, that didn't perform well due to the overhead of bootstrapping an EMR cluster.
+
+Our goal for the this iteration aims to keep some Spark cluster components active at all times. This approach helps minimize cluster bootstrapping durations, but also comes at the price of always having some amount of cost-incurring cluster resources on standby.
+
+## Decision
+
+Given that we are still early in Raster Foundry's product development, the desire to keep costs low and ship quickly carry significant weight. In order to meet those objectives, but still minimize cluster bootstrapping overhead, the first pass at having Spark enabled resources on standby will consist of a shared (across staging and production) Amazon EMR cluster.
+
+The overall Raster Foundry infrastructure will be split across three Amazon Virtual Private Clouds (VPCs):
+
+- Staging (`10.0.0.0/18`, 16382 IPs)
+- Spark (`10.0.64.0/18`, 16382 IPs)
+- Production (`10.0.128.0/18`, 16382 IPs)
+
+The `Staging` and `Production` VPCs will be peered with the `Spark` VPC to allow direct private network access, if necessary. Regardless, public facing APIs such as the Amazon EMR's Step API, or the Raster Foundry API can be used to pass information between systems.
+
+Despite having thought long and hard about the consequences, I suspect that the section below does not cover all of the issues that lie ahead of us. The hope is that at least these core ideas remain in future iterations of the Spark execution environment:
+
+- Shared cluster resources between environments
+- Isolated environments at the VPC level
+- Build on top of the Amazon EMR platform
+- Use spot pricing and auto-scaling aggressively to keep costs low
+
+## Consequences
+
+- All of our current VPCs were configured with the `10.0.0.0/16` CIDR. In order to modify the CIDR ranges, almost all resources housed within the VPC will need to be destroyed and recreated.
+
+- The EMR `Master` instance group requires an on-demand instance, and that instance must be from the `M4`, `C4`, `R3`, `I2`, or `D2` instance families. Reserved Instances will have to be considered to keep costs low.
+
+- Terraform does not yet have complete support for Amazon EMR (some basic support was added in October). In order to take advantage of cost-cutting features like spot bidding (Spot Fleet support hinted at during re:Invent 2016) and auto-scaling for `Core` and `Task` instance groups, we'll need to lean on CloudFormation or direct access to the AWS EMR API.
```

Comment:
Good thought. I believe there is even some AWS provided reference architecture for this pattern with an ECS cluster.

A little hesitant to include it here because I think it is unlikely that we'd take that route. I believe that the Lambda based auto-scaling for ECS approach referenced above was published before "true" service auto-scaling support was added to ECS. Similar "true" support was added to EMR recently, so I think we'd want to take advantage of that.

See:

- https://aws.amazon.com/blogs/compute/automatic-scaling-with-amazon-ecs/
- https://aws.amazon.com/blogs/big-data/dynamically-scale-applications-on-amazon-emr-with-auto-scaling/